### PR TITLE
Add ILPSearch for exact continuous causal discovery

### LINKED ISSUE #3534: [ENH] Finding optimal Causal Structure using the Integer Linear Programming
## Introduction

Learning Directed Acyclic Graphs (DAGs) from continuous observational data is a core task in causal discovery. While heuristic and continuous optimization approaches provide scalable approximations for causal graph learning, exact mathematical programming algorithms offer unique guarantees by finding globally optimal DAG structures accompanied by formal optimality gap certificates.

This proposal aims to introduce an exact Mixed-Integer Linear Programming (MILP) algorithm, `ILPSearch`, into `pgmpy.causal_discovery`. By formulating continuous Linear Structural Equation Models (SEMs) into exact integer programming constraints over directed candidate edges, `ILPSearch` will eliminate the risk of local minima while incorporating prior structural knowledge seamlessly.
By relying on SciPy's built-in MILP solver (`scipy.optimize.milp`), `ILPSearch` provides exact causal structure learning without requiring external solver installations or licensing setup.

---

### Algorithm Steps

The proposed `ILPSearch` algorithm will follow a 5-step pipeline during execution in `_fit(X)`:

1. **Step 1: Empirical Covariance Computation**:
   Compute the sample empirical covariance matrix from input data matrix $\mathcal{X} \in \mathbb{R}^{n \times m}$:
   $$S = \frac{1}{n} \mathcal{X}^T \mathcal{X}$$

2. **Step 2: Superstructure & `ExpertKnowledge` Resolution**:
   If `expert_knowledge` is provided, fit it to resolve candidate search space edges $\mathcal{M} = (V, E)$, required edges, forbidden edges, and temporal orderings. If `expert_knowledge` is None, automatically estimate the moral graph superstructure via Graphical Lasso (sparse inverse covariance estimation).

3. **Step 3: Big-M Parameter Estimation**:
   For each variable $X_k$, run an unconstrained least-squares regression against its candidate parents in $\mathcal{M}$ to find unconstrained coefficients $\hat{\beta}_{jk}^*$. Compute the Big-M bound:

   $$M = 2 \max_{(j,k) \in \vec{E}} \left\vert{} \hat{\beta}_{jk}^* \right\vert{}$$

4. **Step 4: MILP Optimization Model Setup**:
   Construct the Mixed-Integer Linear Program (MILP) matrix and bound vectors for `scipy.optimize.milp`:
   - Define decision variables ($\beta_{jk}, z_{jk}, g_{jk}, \psi_k$).
   - Formulate linear objective and Big-M weight bounds.
   - Add tournament orientation constraints and Layered Network acyclicity constraints.
   - Enforce fixed variable bounds from `required_edges_` ($z_{jk}=1$) and `forbidden_edges_` ($z_{jk}=0$).

5. **Step 5: SciPy Solver Call & Graph Extraction**:
   Invoke `scipy.optimize.milp` to global optimality (or within `opt_gap`). Extract active directed edges ($z_{jk}=1$ and $\beta_{jk} \neq 0$), construct the final `pgmpy.base.DAG` object, and store it in `self.causal_graph_` alongside `self.adjacency_matrix_`.

---

#### 1. Optimization Objective ($\text{PNL}\mathcal{M}$)

For a continuous linear Structural Equation Model (SEM) $X_k = \sum_{j \in \text{pa}(k)} \beta_{jk} X_j + \delta_k$ with sample covariance matrix $S = \frac{1}{n} \mathcal{X}^T \mathcal{X}$, the loss function is:

$$l_n(B) = \frac{1}{2} \text{tr}\left( (I - B)(I - B)^T S \right) = \frac{1}{2n} \sum_{k=1}^m \sum_{d=1}^n \left( x_{dk} - \sum_{(j,k) \in \vec{E}} \beta_{jk} x_{dj} \right)^2$$

The overall optimization problem minimizes the loss plus a sparsity penalty over the candidate directed edge set $\vec{E}$:

$$\min_{B} \quad l_n(B) + \lambda \phi(B) \quad \text{s.t. } G(B) \text{ is an induced DAG from } \overrightarrow{\mathcal{M}}$$

##### $\ell_0$-Regularization Model ($\ell_0\text{-LN}$):
$$
\begin{aligned}
\min_{\beta, z, g, \psi} \quad & \frac{1}{2} \text{tr}\left( (I - B)(I - B)^T S \right) + \lambda \sum_{(j,k) \in \vec{E}} g_{jk} \\
\text{s.t.} \quad & -M g_{jk} \le \beta_{jk} \le M g_{jk}, \quad \forall (j,k) \in \vec{E} && \text{(Big-}M\text{ Weight Bound)} \\
& g_{jk} \le z_{jk}, \quad \forall (j,k) \in \vec{E} && \text{(Active Edge Link)} \\
& z_{jk} + z_{kj} = 1, \quad \forall (j,k) \in \vec{E}, \; j < k && \text{(Tournament Choice)} \\
& z_{jk} - (m-1) z_{kj} \le \psi_k - \psi_j, \quad \forall (j,k) \in \vec{E} && \text{(Layer Acyclicity)} \\
& 1 \le \psi_k \le m, \quad \forall k \in V && \text{(Layer Bounds)} \\
& z_{jk}, g_{jk} \in \{0, 1\}, \quad \beta_{jk} \in \mathbb{R}, \quad \psi_k \in \mathbb{R}
\end{aligned}
$$

##### $\ell_1$-Regularization Model ($\ell_1\text{-LN}$):
$$
\begin{aligned}
\min_{\beta, z, \psi} \quad & \frac{1}{2} \text{tr}\left( (I - B)(I - B)^T S \right) + \lambda \sum_{(j,k) \in \vec{E}} |\beta_{jk}| \\
\text{s.t.} \quad & -M z_{jk} \le \beta_{jk} \le M z_{jk}, \quad \forall (j,k) \in \vec{E} && \text{(Big-}M\text{ Weight Bound)} \\
& z_{jk} + z_{kj} = 1, \quad \forall (j,k) \in \vec{E}, \; j < k && \text{(Tournament Choice)} \\
& z_{jk} - (m-1) z_{kj} \le \psi_k - \psi_j, \quad \forall (j,k) \in \vec{E} && \text{(Layer Acyclicity)} \\
& 1 \le \psi_k \le m, \quad \forall k \in V && \text{(Layer Bounds)} \\
& z_{jk} \in \{0, 1\}, \quad \beta_{jk} \in \mathbb{R}, \quad \psi_k \in \mathbb{R}
\end{aligned}
$$

#### 3. Logical Mapping for `ExpertKnowledge` Constraints

- **Search Space ($\mathcal{M}$)**: Includes only variables $z_{jk}, \beta_{jk}$ for pairs $(j,k) \in \mathtt{search\_space}$.
- **Forbidden Edges**: Adds hard constraints $z_{jk} = 0, \beta_{jk} = 0$.
- **Required Edges**: Adds hard constraints $z_{jk} = 1$.
- **Temporal Tier Ordering**: For node $j$ in tier $T(j)$ and node $k$ in tier $T(k)$ where $T(j) < T(k)$, adds $\psi_j + 1 \le \psi_k$ and $z_{kj} = 0$.

---

## Alternative Solutions

1. **Cutting Plane (CP) Approach**:
   - Uses binary edge variables and dynamically adds cycle-elimination constraints whenever cycles are detected during solver execution (via DFS callbacks).
   - *Comparison*: Requires custom solver callbacks which are not supported by standard `scipy.optimize.milp`.

2. **Linear Ordering (LO) Approach**:
   - Enforces a linear ordering of nodes using binary pairwise ordering variables and 3-cycle elimination constraints.
   - *Comparison*: Generates $O(m^3)$ constraints for $m$ nodes, which slows down branch-and-bound solves.

3. **Topological Ordering (TO) Approach**:
   - Uses a binary permutation matrix to assign topological positions to nodes.
   - *Comparison*: Suffers from high symmetry (multiple topological orderings represent the exact same DAG), causing redundant search tree branches.

---

#### Implementation Workflow

1. **Process Superstructure & `ExpertKnowledge`**:
   - If `expert_knowledge` is provided, call `expert_knowledge.fit(X)` to extract `search_space_`, `required_edges_`, `forbidden_edges_`, and temporal orderings.
   - If `expert_knowledge` is None, estimate candidate superstructure edges via `GraphicalLassoCV` or default to complete graph.

2. **Estimate Big-M Bounds**:
   - Run OLS regressions for each node against candidate parents to estimate raw coefficients.
   - Compute Big-M bound `M = max(2.0 * max_beta, 10.0)` for decision variable bounds.

3. **Construct `scipy.optimize.milp` Inputs**:
   - Concatenate decision vector `x = [z, beta, g, psi]` (binary edge indicators `z`, continuous weights `beta`, $L_0$ penalty flags `g`, layer variables `psi`).
   - Define objective vector `c` with penalty `l_penalty` and `integrality` array.
   - Set `Bounds(lb, ub)` for variables and lock `required_edges_` (`lb=1`) and `forbidden_edges_` (`ub=0`).
   - Build sparse `LinearConstraint(A, clb, cub)` enforcing Big-M bounds, tournament choice, layer acyclicity, and temporal tier ordering.

4. **Execute Solver & Extract DAG**:
   - Solve the MILP model using SciPy's `milp` function.
   - Extract active directed edges into a `DAG`.

### Class API Design

```python
import pandas as pd
from pgmpy.causal_discovery import ExpertKnowledge
from pgmpy.causal_discovery._base import BaseCausalDiscovery

class ILPSearch(BaseCausalDiscovery):
    """
    Exact score-based causal discovery for continuous data using Integer Programming.

    Implements the Layered Network (LN) formulation to find globally optimal DAGs
    with L0 or L1 sparsity penalties using scipy.optimize.milp.

    Parameters
    ----------
    penalty : {'l0', 'l1'}, default='l0'
        Regularization penalty type.

    l_penalty : float, default=0.1
        Sparsity penalty coefficient (lambda).

    expert_knowledge : ExpertKnowledge instance or None, default=None
        Prior expert knowledge specifying search space (superstructure graph),
        forbidden edges, required edges, or temporal ordering. If None, a moral
        graph is automatically estimated from data.

    opt_gap : float, default=0.001
        Target relative MIP optimality gap tolerance.
    """

    def __init__(
        self,
        penalty="l0",
        l_penalty=0.1,
        expert_knowledge=None,
        opt_gap=0.001,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.penalty = penalty
        self.l_penalty = l_penalty
        self.expert_knowledge = expert_knowledge
        self.opt_gap = opt_gap

    def _fit(self, X: pd.DataFrame):
        """
        The fitting procedure for the ILPSearch algorithm.

        Parameters
        ----------
        X : pd.DataFrame
            The data to learn the causal structure from.

        Returns
        -------
        self : ILPSearch
            Fitted instance with learned graph stored in `self.causal_graph_`.
        """
        # 1. Resolve ExpertKnowledge search space, required/forbidden edges
        # 2. Map Layered Network problem to scipy.optimize.milp
        # 3. Solve using SciPy HiGHS backend
        # 4. Store fitted DAG in self.causal_graph_
        return self
```

### Zero-Extra-Dependency Integration

Since `scipy` is already a core dependency of `pgmpy`, `ILPSearch` relies directly on `scipy.optimize.milp` to execute exact integer programming out of the box.

---

## Usage

```python
import pandas as pd
from pgmpy.causal_discovery import ILPSearch, ExpertKnowledge

# 1. Initialize ILP causal discovery algorithm with L0 sparsity penalty
ilp = ILPSearch(penalty="l0", l_penalty=0.1)

# Fit globally optimal DAG from continuous data
ilp.fit(data)

# Access fitted DAG and learned adjacency matrix directly
ilp.causal_graph_
ilp.adjacency_matrix_

# 2. Fit ILP search with prior Expert Knowledge (superstructure & edge constraints)
ek = ExpertKnowledge(
    search_space=[("A", "B"), ("B", "C"), ("A", "C")],
    required_edges=[("A", "B")],
    forbidden_edges=[("C", "A")]
)
ilp_ek = ILPSearch(penalty="l1", l_penalty=0.05, expert_knowledge=ek)
ilp_ek.fit(data)
ilp_ek.causal_graph_
```

---

## References

[1] [Integer Linear Programming for the Bayesian network structure learning problem.](https://linkinghub.elsevier.com/retrieve/pii/S0004370215000417)

[2] [Integer Programming for Learning Directed Acyclic Graphs from Continuous Data](https://pubsonline.informs.org/doi/epdf/10.1287/ijoo.2019.0040)

[3] [Branch and Bound](https://en.wikipedia.org/w/index.php?title=Branch_and_bound&oldid=1352876192)

[4] [Branch and Cut](https://en.wikipedia.org/w/index.php?title=Branch_and_cut&oldid=1284905949)

[5] [Cutting Plane Method](https://en.wikipedia.org/wiki/Cutting-plane_method)

[6] [SciPy `scipy.optimize.milp` Documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.milp.html)

### PR DESCRIPTION
**Pull request by Jatinbhardwaj-093**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.

I used Gemini 3.6 Flash. I utilized two variants: the medium model for fast changes, and the more thoughtful high model for the algorithm review.

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

I used AI for three things: First, I used it to research the algorithm itself. After clearing up all my doubts about the algorithm, I started working on understanding the Scipy Optimizer, slowly building and understanding the vector I needed to construct for it. Secondly, I used AI to check parameters build like objective_vector, LinearConstraints, and Bounds, verifying them with examples. Thirdly, I used AI to determine the best potential data type for the algorithm.

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

I went through the algorithm step by step, rather than having the AI build the entire class at once. I started with a first prototype and then developed each parameter individually, maintaining full control throughout the process. Finally, I verified the implementation by running test cases against the class.

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

I used AI to build a synthetic 5-node graph, which I then used for a test case.

### Issue number(s) that this pull request fixes
- Closes #3534

### List of changes to the codebase in this pull request
- Implement ILPSearch class in pgmpy.causal_discovery for exact score-based causal structure learning via Integer Linear Programming (ILP).
- Add test suite in pgmpy/tests/test_causal_discovery/test_ILPSearch.py (cancer dataset + 4-node graph verification).
- Add ILPSearch to docs/api/structure_learning.rst and citation to docs/references.bib.

### CODE REVIEW

File: pgmpy/causal_discovery/ILPSearch.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    --------
+    >>> import pandas as pd
+    >>> from pgmpy.causal_discovery import ExpertKnowledge, ILPSearch
+    >>> df = pd.DataFrame({"A": [1.0, 2.0, 3.0], "B": [2.0, 4.0, 6.0]})
+    >>> ek = ExpertKnowledge(required_edges=[("A", "B")])
+    >>> ilp = ILPSearch(expert_knowledge=ek, options={"time_limit": 10.0})
+    >>> ilp = ilp.fit(df)
+    >>> ("A", "B") in ilp.causal_graph_.edges()
+    True
+    """
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py:180]: Add a References sections.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    Attributes
+    ----------
+    causal_graph_ : pgmpy.base.DAG or pgmpy.base.PDAG
+        The learned optimal causal graph as a DAG or PDAG instance.
+
+    adjacency_matrix_ : pd.DataFrame
+        Binary adjacency matrix representation of ``causal_graph_``.
+
+    variables_ : list
+        The list of variable names in the dataset used to learn the causal graph.
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: Is this same as `feature_names_in_`? If yes, should remove it.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        ----------
+        X : pd.DataFrame
+            Observational dataset with continuous columns.
+
+        Returns
+        -------
+        self : ILPSearch
+            Fitted estimator with attributes ``causal_graph_`` and ``adjacency_matrix_`` set.
+        """
+        self.variables_ = list(X.columns)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: Shouldn't do this. Use `self.feature_names_in_` instead of `self.variables_`

File: pgmpy/causal_discovery/ILPSearch.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        m = len(self.variables_)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: `m` is same as `self.n_features_in_`. You can directly use that.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+        variable_map = {name: idx for idx, name in enumerate(self.variables_)}
+
+        # Step 1: Superstructure & Expert Knowledge Resolution via CI-test screening
+        if self.expert_knowledge is None:
+            ek = ExpertKnowledge(search_space="marginally_dependent")
+            ek.fit(X)
+            candidate_pairs = set(ek.search_space_)
+            required_edges = set()
+            forbidden_edges = set()
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py:222]: Move this after the `self.expert_knowledge is None` if-else block. Reduces duplicate code. Lines 168-169 can be removed as well.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            ek = cast(ExpertKnowledge, clone(self.expert_knowledge))
+            ek.fit(X)
+            required_edges = set(ek.required_edges_)
+            forbidden_edges = set(ek.forbidden_edges_)
+            if self.expert_knowledge.search_space:
+                candidate_pairs = (set(ek.search_space_) | required_edges) - forbidden_edges
+            else:
+                ek_screen = ExpertKnowledge(search_space="marginally_dependent")
+                ek_screen.fit(X)
+                candidate_pairs = (set(ek_screen.search_space_) | required_edges) - forbidden_edges
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: There are a few repeated lines of code in this block. This of ways to simplify it.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+        # Build candidate directed edge list
+        directed_edges = [(variable_map[u], variable_map[v]) for u, v in candidate_pairs if u != v]
+        num_directed_edges = len(directed_edges)
+        directed_edge_index = {edge: i for i, edge in enumerate(directed_edges)}
+
+        # Step 2: Big-M Estimation via unconstrained OLS
+        X_mat_raw = X.to_numpy(dtype=float)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py:246]: Add a check if there is any categorical variables in `X`. If that is the case, we should throw an error.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+        # Layout: x = [z (num_directed_edges), beta (num_directed_edges), g (num_directed_edges if L0), psi (m)]
+        has_g = self.penalty == "l0"
+        offset_z = 0
+        offset_beta = offset_z + num_directed_edges
+        offset_g = offset_beta + num_directed_edges if has_g else offset_beta
+        offset_psi = (offset_g + num_directed_edges) if has_g else (offset_beta + num_directed_edges)
+        n_solver_vars = offset_psi + m
+
+        # Step 4: Objective Vector c
+        c = np.zeros(n_solver_vars)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: Use better variable names.

File: pgmpy/causal_discovery/ILPSearch.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+                row = np.zeros(n_solver_vars)
+                row[offset_z + directed_edge_index[(j, k)]] = 1
+                row[offset_z + directed_edge_index[(k, j)]] = -(m - 1)
+                row[offset_psi + j] = 1
+                row[offset_psi + k] = -1
+                A_rows.append(row)
+                clb.append(-np.inf)
+                cub.append(0.0)
+
+        A_mat = csc_matrix(A_rows) if A_rows else csc_matrix((0, n_solver_vars))
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: Use a better name here as well.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by Jatinbhardwaj-093** [comment]
Screenshot 2026-08-12 at 8 45 27 PM
Screenshot 2026-08-12 at 8 46 00 PM
Screenshot 2026-08-12 at 8 46 12 PM
Screenshot 2026-08-12 at 8 46 17 PM

**Comment by Jatinbhardwaj-093** [comment]
[ILP_paper_script.py](https://github.com/user-attachments/files/30987388/ILP_paper.py)
This script handles all configurations and evaluation metrics, I also added the FNR. It covers 8 permutations for m={10, 20, 30, 40} and n={100, 1000}, testing both Moral and complete graphs with l_penalty values of {0.1, 1}. 
There is an issue with the case using an L0 penalty with a lambda of 1. The optimal LB and UB reach 0, resulting in nearly empty graphs.
I will continue work on this to achieve results more consistent with the paper.

**Comment by Jatinbhardwaj-093** [comment]
Screenshot 2026-08-19 at 4 21 23 PM
Screenshot 2026-08-19 at 4 21 38 PM

File: pgmpy/causal_discovery/ILPSearch.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+        Returns
+        -------
+        self : ILPSearch
+            Fitted estimator with attributes ``causal_graph_`` and ``adjacency_matrix_`` set.
+        """
+        # Step 0: Continuous Data Validation
+        try:
+            X.astype(float)
+        except (ValueError, TypeError):
+            raise ValueError("ILPSearch requires continuous (numeric) variables.")
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ILPSearch.py]: Use `pgmpy.utils.get_dataset_type` instead.

**Comment by Jatinbhardwaj-093** [comment]
Hello @ankurankan, could you please review this when you have time?

Changes Made:
* As suggested, I improved the step comments and the comments on the parameters and variables for this algorithm. I also renamed a few variables.
* Removed the L1 regularizer and added proper documentation noting why this can't be implemented in the MILP solver.
* Removed the default marginal independence search space. Now, if a search space is not provided, we run the optimizer on a complete graph.
* Improved the docs and added a proper example showing all attributes of the class.
* Added a new synthetic data graph with three layers and one collider at layer 2.
	* Previously, the synthetic graph I created produced a very dense graph because a very low `l_penalty=0.001` was used for discovery.
* Added tests for the Sachs dataset.
* Added test cases for different parameters of expert knowledge for rand_data, Cancer, Sachs.

**Comment by Jatinbhardwaj-093** [comment]
Screenshot 2026-08-24 at 6 02 56 PM


In the build docs, there are two references for the paper. Why is this happening?
